src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs

```diff
@@ -69,47 +73,115 @@ IAsyncServiceProvider serviceProvider
             IProjectSystemCache projectSystemCache,
             ISolutionRestoreWorker restoreWorker,
             ILogger logger,
-            Microsoft.VisualStudio.Threading.AsyncLazy<IVsSolution2> vsSolution2)
+            Microsoft.VisualStudio.Threading.AsyncLazy<IVsSolution2> vsSolution2,
+            INuGetTelemetryProvider telemetryProvider)
         {
             _projectSystemCache = projectSystemCache ?? throw new ArgumentNullException(nameof(projectSystemCache));
             _restoreWorker = restoreWorker ?? throw new ArgumentNullException(nameof(restoreWorker));
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
             _vsSolution2 = vsSolution2 ?? throw new ArgumentNullException(nameof(vsSolution2));
+            _telemetryProvider = telemetryProvider ?? throw new ArgumentNullException(nameof(telemetryProvider));
         }
 
-        public Task<bool> CurrentRestoreOperation => _restoreWorker.CurrentRestoreOperation;
+        Task<bool> IVsSolutionRestoreService.CurrentRestoreOperation
+        {
+            get
+            {
+                const string eventName = nameof(IVsSolutionRestoreService) + "." + nameof(IVsSolutionRestoreService.CurrentRestoreOperation);
+                NuGetExtensibilityEtw.EventSource.Write(eventName, NuGetExtensibilityEtw.InfoEventOptions);
+                return _restoreWorker.CurrentRestoreOperation;
+            }
+        }
 
-        public async Task<bool> NominateProjectAsync(string projectUniqueName, CancellationToken token)
+        Task<bool> IVsSolutionRestoreService3.CurrentRestoreOperation
         {
-            Assumes.NotNullOrEmpty(projectUniqueName);
+            get
+            {
+                const string eventName = nameof(IVsSolutionRestoreService) + "." + nameof(IVsSolutionRestoreService3.CurrentRestoreOperation);
+                NuGetExtensibilityEtw.EventSource.Write(eventName, NuGetExtensibilityEtw.InfoEventOptions);
+                return _restoreWorker.CurrentRestoreOperation;
+            }
+        }
 
-            ProjectNames projectNames = await GetProjectNamesAsync(projectUniqueName, token);
-            var dgSpec = new DependencyGraphSpec();
-            var packageSpec = new PackageSpec()
+        public async Task<bool> NominateProjectAsync(string projectUniqueName, CancellationToken token)
+        {
+            const string eventName = nameof(IVsSolutionRestoreService2) + "." + nameof(NominateProjectAsync);
```

Comment:
I know this might be asking too much, but can this logic be taken *out* of the business logic here?
Partial class is an option, but maybe having another interface that calls this one would be better?

Reply:
What do you think of this?

```cs
[Export(typeof(IVsPackageInstaller))]
internal class VsPackageInstallerExported
{
  private IVsPackageInstaller _inner;

  [ImportingConstructor]
  public VsPackageInstallerExported(args)
  {
    _inner = new VsPackageInstaller(args);
  }

  public void InstallPackage(Project project, string packageId, string packageVersion)
  {
    const string eventName = "...";
    NuGetEtw.EventSource.Write(eventName, NuGetEtw.StartOptions);
   try
   {
     _inner.InstallPackage(project, packageId, packageVersion);
   }
   finally
   {
     NuGetETw.EventSource.Write(eventName, NuGetEtw.StopOptions);
   }
  }
}
```

and also:

```diff
-[Export(typeof(VsPackageInstaller))]
 public VsPackageInstaller : IVsPackageInstaller
 {
-  [ImportingConstructor]
   public VsPackageInstaller(args)
   {
     // whatever
   }
 }
```

Reply:
Yeah that was basically my suggestion. 

There's obviously a trade off there (with the exports and all), so I'm wondering if it might be too much. 

The IDisposables *might* be cleaner/add less long term overhead. 

Alternatively, this would allow you to enforce everything is instrumented.

Reply:
Can we use annotations similar to [this](https://docs.microsoft.com/en-us/dotnet/core/extensions/logger-message-generator)? Code will look much cleaner, of course it might not fit everywhere. If I remember correctly, we can have both before execution and after execution attribution runs.

Reply:
@erdembayar I was looking into writing custom source generators, but as I don't have any experience or something similar to copy, I couldn't get anything completed within a few hours. I agree that using attributes and generating a lot of this code would be great, but it's too much effort for me to do.